File: src/collections_and_containers/cpp/linked_list/src/linked_list.cpp
#include "linked_list.h"
#include <stdexcept>

template <class T>
LinkedList<T>::LinkedList() : head(nullptr), tail(nullptr), n(0) {}

template <class T>
LinkedList<T>::LinkedList(const LinkedList<T> &l)
    : head(nullptr), tail(nullptr), n(0) {
  const Node *p = l.head.get();
  while (p != nullptr) {
    push_back(p->data);
    p = p->next.get();
  }
}

template <class T>
LinkedList<T> &LinkedList<T>::operator=(const LinkedList<T> &l) {
  if (this != &l) {
    clear();
    const Node *p = l.head.get();
    while (p != nullptr) {
      push_back(p->data);
      p = p->next.get();
    }
  }
  return *this;
}

template <class T> LinkedList<T>::~LinkedList() { clear(); }

template <class T> void LinkedList<T>::push_back(const T &v) {
  auto newNode = std::make_unique<Node>();
  newNode->data = v;
  
  if (empty()) {
    tail = newNode.get();
    head = std::move(newNode);
  } else {
    tail->next = std::move(newNode);
    tail = tail->next.get();
  }
  n++;
}

template <class T> void LinkedList<T>::push_front(const T &v) {
  auto newNode = std::make_unique<Node>();
  newNode->data = v;
  
  if (empty()) {
    tail = newNode.get();
    head = std::move(newNode);
  } else {
    newNode->next = std::move(head);
    head = std::move(newNode);
  }
  n++;
}

template <class T> void LinkedList<T>::pop_back() {
  if (empty()) {
    throw std::out_of_range("pop_back() called on empty list");
  }
  if (n == 1) {
    head = nullptr;
    tail = nullptr;
  } else {
    Node *p = head.get();
    while (p->next.get() != tail) {
      p = p->next.get();
    }
    p->next = nullptr;
    tail = p;
  }
  n--;
}

template <class T> void LinkedList<T>::pop_front() {
  if (empty()) {
    throw std::out_of_range("pop_front() called on empty list");
  }
  if (n == 1) {
    head = nullptr;
    tail = nullptr;
  } else {
    head = std::move(head->next);
  }
  n--;
}

template <class T> T &LinkedList<T>::front() {
  if (empty()) {
    throw std::out_of_range("front() called on empty list");
  }
  return head->data;
}

template <class T> const T &LinkedList<T>::front() const {
  if (empty()) {
    throw std::out_of_range("front() called on empty list");
  }
  return head->data;
}

template <class T> T &LinkedList<T>::back() {
  if (empty()) {
    throw std::out_of_range("back() called on empty list");
  }
  return tail->data;
}

template <class T> const T &LinkedList<T>::back() const {
  if (empty()) {
    throw std::out_of_range("back() called on empty list");
  }
  return tail->data;
}

template <class T> bool LinkedList<T>::empty() const { return n == 0; }

template <class T> void LinkedList<T>::clear() {
  head = nullptr;
  tail = nullptr;
  n = 0;
}

template <class T> int LinkedList<T>::size() const { return n; }

template <class T> void LinkedList<T>::remove(const T &v) {
  if (empty()) {
    throw std::out_of_range("remove() called on empty list");
  }
  if (head->data == v) {
    pop_front();
    return;
  }
  
  Node *p = head.get();
  while (p->next != nullptr && p->next->data != v) {
    p = p->next.get();
  }
  
  if (p->next == nullptr) {
    throw std::out_of_range("remove() called on non-existent value");
  }
  
  if (p->next.get() == tail) {
    tail = p;
    p->next = nullptr;
    n--;
  } else {
    p->next = std::move(p->next->next);
    n--;
  }
}

template <class T> bool LinkedList<T>::contains(const T &v) const {
  const Node *p = head.get();
  while (p != nullptr) {
    if (p->data == v) {
      return true;
    }
    p = p->next.get();
  }
  return false;
}

template class LinkedList<int>;
template class LinkedList<double>;
template class LinkedList<std::string>;

File: src/collections_and_containers/python/linked_list/src/linked_list.py
"""Singly linked list implementation."""

from __future__ import annotations

from typing import Any, Iterator, Optional


class Node:
    """A node in a singly linked list."""

    def __init__(self, data: Any) -> None:
        self.data = data
        self.next: Optional[Node] = None


class LinkedList:
    """A singly linked list data structure."""

    def __init__(self) -> None:
        self.head: Optional[Node] = None
        self.n = 0

    def push_back(self, v: Any) -> None:
        """Add an element to the end of the list."""
        if self.head is None:
            self.head = Node(v)
        else:
            node = self.head
            while node.next is not None:
                node = node.next
            node.next = Node(v)
        self.n += 1

    def push_front(self, v: Any) -> None:
        """Add an element to the front of the list."""
        node = Node(v)
        node.next = self.head
        self.head = node
        self.n += 1

    def pop_back(self) -> Optional[Any]:
        """Remove and return the last element."""
        if self.head is None:
            return None

        last = self.head
        prev = None
        while last.next is not None:
            prev = last
            last = last.next

        if last == self.head:
            self.head = None
        else:
            prev.next = None

        self.n -= 1
        return last.data

    def pop_front(self) -> Optional[Any]:
        """Remove and return the first element."""
        if self.head is None:
            return None

        node = self.head
        self.head = node.next
        self.n -= 1
        return node.data

    def front(self) -> Optional[Any]:
        """Return the first element without removing it."""
        if self.head is None:
            return None
        return self.head.data

    def back(self) -> Optional[Any]:
        """Return the last element without removing it."""
        if self.head is None:
            return None

        node = self.head
        while node.next is not None:
            node = node.next
        return node.data

    def empty(self) -> bool:
        """Check if the list is empty."""
        return self.head is None

    def clear(self) -> None:
        """Remove all elements from the list."""
        self.head = None
        self.n = 0

    def size(self) -> int:
        """Return the number of elements in the list."""
        return self.n

    def remove(self, v: Any) -> None:
        """Remove the first occurrence of an element."""
        if self.head is None:
            return

        if self.head.data == v:
            self.head = self.head.next
            self.n -= 1
            return

        node = self.head
        while node.next is not None:
            if node.next.data == v:
                node.next = node.next.next
                self.n -= 1
                return
            node = node.next

    def __iter__(self) -> Iterator[Any]:
        node = self.head
        while node is not None:
            yield node.data
            node = node.next

    def __str__(self) -> str:
        return str(list(self))

    def __repr__(self) -> str:
        return str(self)
